apps/server/src/modules/files-storage/client/s3-client.adapter.ts

```diff
@@ -189,37 +187,43 @@ export class S3ClientAdapter implements IStorageClient {
 		return this.client.send(req);
 	}
 
-	public async list(prefix: string, maxKeys?: number) {
-		this.logger.log({ action: 'list', params: { prefix, bucket: this.config.bucket } });
-
+	public async list(params: IListFiles) {
 		try {
-			let files: string[] = [];
-			let ret: ListObjectsV2CommandOutput | undefined;
+			this.logger.log({ action: 'list', params });
 
-			do {
-				const req = new ListObjectsV2Command({
-					Bucket: this.config.bucket,
-					Prefix: prefix,
-					ContinuationToken: ret?.NextContinuationToken,
-					MaxKeys: maxKeys && maxKeys - files.length,
-				});
+			const result = await this.listObjectKeysRecursive(params);
+
+			return result;
+		} catch (err) {
+			throw new InternalServerErrorException('S3ClientAdapter:listDirectory');
+		}
+	}
 
-				// Iterations are dependent on each other
-				// eslint-disable-next-line no-await-in-loop
-				ret = await this.client.send(req);
+	private async listObjectKeysRecursive(params: IListFiles) {
+		const { path, maxKeys, nextMarker } = params;
+		const files: string[] = params.files ? params.files : [];
 
-				const returnedFiles =
-					ret?.Contents?.filter((o) => o.Key)
-						.map((o) => o.Key as string) // Can not be undefined because of filter above
-						.map((key) => key.substring(prefix.length)) ?? [];
+		const req = new ListObjectsV2Command({
+			Bucket: this.config.bucket,
+			Prefix: path,
+			ContinuationToken: nextMarker,
+			MaxKeys: maxKeys,
```

Comment:
Just passing in `maxKeys` can lead to situations where we return more files than requested (e.g. passing 1500 as a value for maxKeys will return up to 2000 filenames)

Would need to be something like
```typescript
MaxKeys: maxKeys && (maxKeys - files.length),
```

Reply:
You are sure? For me it make not really sense. We pass the marker (nextMarker) that hopefully work like a skip. We pass the MayKeys that should work like a limit. Otherwise the marker and MaxKeys make no sense and must always self handled implemented.

1501 exists and max=1200 is passed
first request 1000
if (true & true) for the condtion check if it call more
second request ---> should start at 1001 and know it, it also know that it only should get 1200 and should return 200.

Reply:
@CeEv it is correctly, we need it!

Reply:
@CeEv testing reveals that MaxKeys is strictly a limit on the returned files per request, not in total. So I think it is needed (Or just fetch files in 1000 increments and slice the array at the end)